File: application/backend-credit-card-enrollment/backend-java/src/main/java/cloud/ambar/creditcard/enrollment/projection/enrollmentlist/EnrollmentRepository.java
package cloud.ambar.creditcard.enrollment.projection.enrollmentlist;

import cloud.ambar.common.projection.MongoTransactionalProjectionOperator;
import lombok.RequiredArgsConstructor;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.stereotype.Service;
import org.springframework.web.context.annotation.RequestScope;

import java.util.List;
import java.util.Optional;

@RequestScope
@Service
@RequiredArgsConstructor
public class EnrollmentRepository {
    private final MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator;

    public void save(final Enrollment enrollment) {
        mongoTransactionalProjectionOperator.operate().save(enrollment, "CreditCard_Enrollment_Enrollment");
    }

    public Optional<Enrollment> findOneById(final String id) {
        return Optional.ofNullable(mongoTransactionalProjectionOperator.operate().findOne(
                Query.query(
                        Criteria.where("id").is(id)
                ),
                Enrollment.class,
                "CreditCard_Enrollment_Enrollment"
        ));
    }

    public List<Enrollment> findAllByUserId(final String userId) {
        return mongoTransactionalProjectionOperator.operate().find(
                Query.query(
                        Criteria.where("userId").is(userId)
                ),
                Enrollment.class,
                "CreditCard_Enrollment_Enrollment"
        );
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/CreditCard/Enrollment/Projection/EnrollmentList/EnrollmentRepository.cs
using CreditCardEnrollment.Common.Projection;
using MongoDB.Driver;

namespace CreditCardEnrollment.CreditCard.Enrollment.Projection.EnrollmentList;

public class EnrollmentRepository {
    private readonly MongoTransactionalProjectionOperator _mongoOperator;

    public EnrollmentRepository(MongoTransactionalProjectionOperator mongoOperator) {
        _mongoOperator = mongoOperator;
    }

    public void Save(Enrollment enrollment) {
        Collection().ReplaceOne(
            e => e.Id == enrollment.Id,
            enrollment,
            new ReplaceOptions { IsUpsert = true }
        );
    }

    public Enrollment? FindOneById(string id) {
        return Collection().Find(e => e.Id == id).FirstOrDefault();
    }

    public IEnumerable<Enrollment> FindAllByUserId(string userId) {
        return Collection().Find(e => e.UserId == userId).ToEnumerable();
    }

    private IMongoCollection<Enrollment> Collection() {
        return _mongoOperator.Operate().GetCollection<Enrollment>("CreditCard_Enrollment_Enrollment");
    }
}